src/ServiceControl/Operations/BodyStorage/RavenAttachments/GetBodyById.cs

```diff
@@ -2,20 +2,65 @@
 {
     using System;
     using System.IO;
+    using System.Linq;
+    using System.Net.Http;
     using CompositeViews.Messages;
     using System.Text;
     using Nancy;
+    using NServiceBus.Logging;
     using Raven.Abstractions.Data;
     using Raven.Client;
     using ServiceBus.Management.Infrastructure.Nancy.Modules;
+    using ServiceControl.Infrastructure.Settings;
 
     public class GetBodyById : BaseModule
     {
+        static ILog logger = LogManager.GetLogger<GetBodyById>();
+        public Func<HttpClient> HttpClientFactory { get; set; }
 
         public GetBodyById()
         {
             Get["/messages/{id*}/body", true] = async (parameters, token) =>
             {
+                var query = (DynamicDictionary)Request.Query;
+
+                var localInstanceId = InstanceIdGenerator.FromApiUrl(Settings.ApiUrl);
+
+                dynamic instanceId;
+                if (query.TryGetValue("instance_id", out instanceId) && instanceId != localInstanceId)
+                {
+                    var remoteUri = InstanceIdGenerator.ToApiUrl(instanceId);
+                    var instanceUri = new Uri($"{remoteUri}{Request.Path}?{Request.Url.Query}");
+
+                    var httpClient = HttpClientFactory();
+                    try
+                    {
+                        var rawResponse = await httpClient.SendAsync(new HttpRequestMessage(HttpMethod.Get, instanceUri)).ConfigureAwait(false);
+
+                        if (rawResponse.StatusCode == System.Net.HttpStatusCode.OK)
+                        {
+                            return new Response
+                                {
+                                    Contents = stream => rawResponse.Content.CopyToAsync(stream).GetAwaiter().GetResult()
+                                }
+                                .WithContentType(rawResponse.Content.Headers.ContentType.ToString())
+                                .WithHeader("Expires", DateTime.UtcNow.AddYears(1).ToUniversalTime().ToString("R"))
+                                .WithHeader("Content-Length", rawResponse.Content.Headers.ContentLength.ToString())
```

Comment:
Can we just iterate through the remote headers and return all of them? If we can generically copy the http response into a nancy response then we can abstract a lot of this away. 

On incoming, if the request is for a remote then call it and pass through it's response.
If the request is not for a remote then execute the local code and return a response.

Reply:
> I'd want to see a lot of the mechanics removed from the Nancy module

@mikeminutillo do you mean moved out from Nancy module into a "helper" class? Yes, that would make sense to keep the module simpler. Keep in mind that we've just added code to delegate to remotes to what was already in place.

Reply:
👍 for making a copy of the selected headers we care about.